File: 75.sort-colors.js
/*
 * @lc app=leetcode id=75 lang=javascript
 *
 * [75] Sort Colors
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
const sortColors = function (nums) 
{
    const length = nums.length;
    if (length >= 2)
    {
        let left = 0;
        let right = length - 1;
        for (let i = 0; i <= right; i++)
        {
            if (nums[i] === 0)
            {
                if (i === left)
                {
                    continue;
                }
                swap(nums, i, left);
                left++;
                i--;
            }
            if (nums[i] === 2)
            {
                if (i === right)
                {
                    continue;
                }
                swap(nums, i, right);
                right--;
                i--;
            }
        }
    }
};

/**
 * @param {number[]} nums 
 * @param {number} index1 
 * @param {number} index2 
 */
function swap(nums, index1, index2)
{
    [nums[index1], nums[index2]] = [nums[index2], nums[index1]];
}
// @lc code=end

File: src/2025/75.sort-colors.cpp
/*
 * @lc app=leetcode id=75 lang=cpp
 *
 * [75] Sort Colors
 */

#include <array>
#include <vector>

// @lc code=start
class Solution {
 public:
  void sortColors(std::vector<int>& nums) {
    std::array<int, 3> num_counts = {0, 0, 0};
    for (const int num : nums) {
      num_counts[num]++;
    }

    int current_index = 0;
    for (int num = 0; num < 3; num++) {
      while (num_counts[num] > 0) {
        nums[current_index] = num;
        current_index++;
        num_counts[num]--;
      }
    }
  }
};
// @lc code=end
